### PR TITLE
Move Impala backend to backends/ directory

### PR DESCRIPTION
**Pull request by matthewmturner**
Did a first iteration of the move from impala to backends/impala.  Some imports arent resolving, but it seems like that would have been the case even before move, to be discussed.  Still need to work on documentation update as well.

### CODE REVIEW

File: ibis/backends/impala/client.py
Comment (datapythonista) [inline ibis/backends/impala/client.py:227]: I think this `impala` is the `impyla` package one, quite misleading, I know.

File: ibis/backends/impala/compat.py
Comment (datapythonista) [inline ibis/backends/impala/compat.py:17]: Again, this is impyla

File: ibis/backends/impala/compiler.py
Comment (datapythonista) [inline ibis/backends/impala/compiler.py:18]:
```suggestion
from . import identifiers
```

File: setup.cfg
Comment (datapythonista) [inline setup.cfg:19]: I think this is being changed because now you're not importing impyla anymore, when you restore the imports it'll probably be back.

**Review by datapythonista — COMMENTED** [review]
Thanks @matthewmturner, looks good, just some confusion with the ambiguos impala module.

File: ibis/backends/impala/client.py
Comment (matthewmturner) [reply ibis/backends/impala/client.py:227]: thanks for info.  in compat.py there is an import dbapi as impyla but i dont see dbapi.py anywhere.  Is that what you are referring to?

Comment (datapythonista) [reply ibis/backends/impala/client.py:227]: There is an external package (i.e. `pip install impyla`) that provides an impala modules (i.e. `import impala`). That's unrelated to Ibis. Then Ibis has a `ibis.impala` module (now moved to `ibis.backends.impala` in this PR), and also, the ibis module has a `ibis.impala` attribute, that is an alias of `ibis.backends.impala` now (you updated the import in `ibis/__init__.py`).

Very confusing, surely not great, but for now just revert the changes here and in the rest of places I commented. Since you thought this was the `impala` in `ibis.impala`, but it's the `impala` in the `impyla` package.

File: ibis/backends/impala/compat.py
Comment (matthewmturner) [inline ibis/backends/impala/compat.py:18]: are these two errors from impyla as well?

Comment (datapythonista) [reply ibis/backends/impala/compat.py:18]: Yes

File: ibis/backends/impala/client.py
Comment (matthewmturner) [reply ibis/backends/impala/client.py]: okay, i think i get it.  i think i fixed it.  also i highlighted another area that i think is impacted by this (errors highlighted below).

File: setup.cfg
Comment (matthewmturner) [reply setup.cfg]: you were right :)

File: ibis/__init__.py
Comment (matthewmturner) [inline ibis/__init__.py:32]: Tests are failing due to `ModuleNotFoundError: No module named 'ibis.impala'` - but i thought the above would have enabled that and the approach looked the same as OmniSci. am i missing something?

Comment (datapythonista) [reply ibis/__init__.py:32]: There is something weird, and it's that the bigquery backend uses many things from the impala one (doesn't make sense, but there must be a historical reason for it). If you see the error, this is the import that is failing:
```
ibis/bigquery/compiler.py:17: in <module>
    from ibis.impala import compiler as impala_compiler
```
You'll have to update it to be `from ibis.backends.impala...`.

File: ibis/backends/impala/client.py
Comment (matthewmturner) [reply ibis/backends/impala/client.py]: im still getting no module not found error for impala, but i dont see impyla in the environment setup. does it need to be added so we can import it?

Comment (datapythonista) [reply ibis/backends/impala/client.py]: If you mean locally, we have every backend dependencies in `ci/deps/<backend>.yml`.

Comment (matthewmturner) [reply ibis/backends/impala/client.py]: Ok, i see it there. I just didnt see it in the github actions setup environment step or the environment.yml (so it isnt resolving locally either since my conda env is based on that file).

Im a bit new to building packages so im going to look into it more.

Comment (matthewmturner) [reply ibis/backends/impala/client.py]: I saw on another pull request there were ci failures due to ibis dependencies not working. Could it be related?

**Comment by datapythonista** [comment]
@matthewmturner there are some failures in Ibis dependencies, but that's causing only 3 builds to fail.

There is something I don't understand here. I'm checking one of the SQL backends jobs, which is failing because of this:
```
ibis/tests/all/conftest.py:11: in <module>
    from ibis.tests.backends import Backend
ibis/tests/backends.py:17: in <module>
    from ibis.bigquery.tests.conftest import connect as bigquery_connect
ibis/bigquery/__init__.py:1: in <module>
    from ibis.bigquery.udf.api import udf  # noqa: F401
ibis/bigquery/udf/__init__.py:1: in <module>
    from ibis.bigquery.udf.api import udf  # noqa: F401
ibis/bigquery/udf/api.py:9: in <module>
    from ibis.bigquery.compiler import BigQueryUDFNode, compiles
ibis/bigquery/compiler.py:16: in <module>
    from ibis.backends.impala import compile as impala_compiler
ibis/backends/impala/__init__.py:6: in <module>
    from .client import (  # noqa: F401
ibis/backends/impala/client.py:29: in <module>
    from .compat import HS2Error, ImpylaError, impyla
ibis/backends/impala/compat.py:15: in <module>
    from impala import dbapi as impyla  # noqa
E   ModuleNotFoundError: No module named 'impala'
```
I think you just changed that line from `import impala.dbapi as impyla` to `from impala import dbapi as impyla`, which should be equivalent, and shouldn't be the cause.

The cause seems to be that `impyla` is not installed, which is expected, since that is the job of the SQL backends. From the traceback looks like the reason for being imported is an import to the bigquery backend from the tests. There are few things that are not great here (the bigquery backend importing the impala compiler, and the generic tests importing the bigquery backend). But that code is not new in this PR, and it's working in master. So, not sure what's the problem, but apparently something is making that code in impala be imported in your PR, when it's not in master.

**Comment by matthewmturner** [comment]
Thanks for the color.

Just to try I used the other import syntax `import impala.dbapi as impyla` and i also noticed that i needed to update some of the tests in backends/impala.  Same results.

What should be the approach here?  

1.  Try to see why the imports arent happening in master and causing issues?  If so, is there a way for me to check that?
2.  Fix the structural issues (the bigquery backend importing the impala compiler, and the generic tests importing the bigquery backend).  If this approach, would those updates belong in this PR or a separate one that should be done before finishing this one?
3. Anything else...?

For my info, should all github action tests pass with the following logic?

- Pandas / files / Pyspark / Spark / OmniSciDB : pass because backends/impala not imported
- Impala / Clickhouse / SQL : pass with backends/impala imported

**Comment by matthewmturner** [comment]
I'm new to testing but trying to compare my PR to master and see whats causing it to be imported in mine and not in master for the non sql tests.  Is there a specific place to check for where being skipped is configured?

**Comment by datapythonista** [comment]
@matthewmturner let's wait until #2443 is merged. I'm not quite sure what's the cause, but that PR may fix the error, and it's something we want to have anyway.

**Comment by datapythonista** [comment]
@matthewmturner #2443 has been merged. If you can merge master into your branch, may be that fixes the CI.

Also, you'll have to rename the module in the new conda recipe we added, https://github.com/ibis-project/ibis/blob/master/ci/recipe/meta.yaml#L64 to get the CI passing (`ibis.impala` -> `ibis.backends.impala`).

**Comment by matthewmturner** [comment]
Thanks, I merged.  Still getting issues. And it looks like the sql tests are still running even in the pandas / files test.  

`Run PYTEST_BACKENDS=$BACKENDS PYTEST_EXPRESSION="not hdfs" ./ci/run_tests.sh
  PYTEST_BACKENDS=$BACKENDS PYTEST_EXPRESSION="not hdfs" ./ci/run_tests.sh
  shell: /bin/bash --noprofile --norc -e -o pipefail {0}
  env:
    BACKENDS: pandas csv parquet hdf5
TESTS_DIRS: ibis/tests ibis/file/tests ibis/pandas/tests
PYTEST_EXPRESSION: not hdfs`

But then we get errors:

`ERROR collecting ibis/tests/sql/test_compiler.py`
`ERROR collecting ibis/tests/sql/test_sqlalchemy.py`

Is my understanding correct that these tests shouldnt be running?

Then in the sql test it looks like were still missing impyla in the dependencies which i also think isnt expected.

What are your thoughts?

File: ibis/tests/sql/test_compiler.py
Comment (datapythonista) [inline ibis/tests/sql/test_compiler.py:10]: I think those may be related to the errors in the CI. Can you move the imports inside the functions when they are being used, and also create an issue and add a `TODO` to them? SQL backends shouldn't depend on the impala backend.

If you don't mind, probably better to do this change in a separate PR (also with the bigquery one). You'll have a small conflict to fix here later, but shouldn't be difficult to resolve.

**Review by datapythonista — COMMENTED** [review]
Thanks a lot for working on this @matthewmturner. Looks like there are several bad design decisions that are affecting these changes. I don't understand why, since we're not making any change here, more than moving a directory, but we may have to fix these problems before we can move forward. See review comments.

File: ibis/bigquery/compiler.py
Comment (datapythonista) [reply ibis/bigquery/compiler.py:17]: I just saw that you replaces `from ibis.impala import compiler` by `from ibis.backends.impala import compile`. See the ending `r`. This is a typo, `compiler` refers to `compiler.py`, this is probably failing.

**Comment by datapythonista** [comment]
I started to refactor the weird dependencies that impala has in #2449, still more work is needed, but hopefully that will help fix your problems here.

File: ibis/tests/sql/test_compiler.py
Comment (matthewmturner) [reply ibis/tests/sql/test_compiler.py:10]: Sure. Would this be part of #2449 or a new issue?

Comment (matthewmturner) [reply ibis/tests/sql/test_compiler.py:10]: or more generally, are any new issues needed now or will it all be in #2449?

File: ibis/backends/impala/tests/test_udf.py
Comment (matthewmturner) [reply ibis/backends/impala/tests/test_udf.py:10]: I get the following error on this suggestion.

`Relative imports cannot be used with "import .a" form; use "from . import a" instead`

File: ibis/bigquery/compiler.py
Comment (matthewmturner) [reply ibis/bigquery/compiler.py:17]: thx for noticing that - fixed.

File: ibis/tests/sql/test_compiler.py
Comment (datapythonista) [reply ibis/tests/sql/test_compiler.py:10]: Some more work is needed after #2449 is merged. The biggest part is to move the operation registry to the new base SQL backend. You can probably wait until that refactoring is done, things will be easy then. But feel free to apply this suggestion if yoy want to move things forward with this PR, that will take some time.

**Comment by datapythonista** [comment]
You've got several conflicts, sorry for that, I fixed the isort issues and didn't think it'd conflict with the open PRs. You may want to run `isort .` from the root of ibis and commit, before merging master. With some luck you won't have to fix any conflict with that.

**Comment by matthewmturner** [comment]
I have isort setup as a precommit hook so I think it's already run - unless I'm missing something.  Is there another functionality it has from running manually?

**Comment by datapythonista** [comment]
Not sure about the precommit hook, it may only change the updated files. I'd do what I suggested, but you can simply merge master too, but I expect you'll have to fix several conmits manually then.

**Comment by matthewmturner** [comment]
I think i fixed that part...i ran isort, merged master, resolved the conflicts.  CI runs now.  But i suppose until the operation registry is moved we will continue to have errors, is that correct?

Is there anything i can do to help move that forward as it looks like #2449 might be close to being done?

thanks again for your guidance!

**Comment by datapythonista** [comment]
@matthewmturner #2449 has been merged. There is still some work on other backends not depending of impala, but if you want to merge master and see if that PR already helps...

**Comment by matthewmturner** [comment]
I merged with master.  Back to getting import error from impyla. But now its coming from the below instead of bigquery:

```
ibis/tests/test_filesystems.py:12: in <module>
    from ibis.backends.impala.tests.conftest import IbisTestEnv
ibis/backends/impala/__init__.py:6: in <module>
    from .client import (  # noqa: F401
ibis/backends/impala/client.py:29: in <module>
    from .compat import HS2Error, ImpylaError, impyla
ibis/backends/impala/compat.py:15: in <module>
    import impala.dbapi as impyla  # noqa
E   ModuleNotFoundError: No module named 'impala'
```

Should  `IbisTestEnv` be using impala when its doing file system testing?

**Comment by datapythonista** [comment]
> Should `IbisTestEnv` be using impala when its doing file system testing?

In Ibis `filesystems` means actually `hdfs`, and `hdfs` is only supported with Impala afaik. I thought on moving it inside the backend, but it's part of the public API, so users may be using it, so it requires more thought, discussion, and probably a deprecation warning first.

In any case, we should be calling pytest with `not hdfs`, and that should prevent `test_filesystems` from running. But I guess that import is in the global namespace, and that's causing the issue. Not sure why this is just failing in this PR. If you want to move that import inside the test/tests where it's being used (assuming it's not too many) that would be good. Probably better in a separate PR.

Thanks for all the work with this @matthewmturner, really happy to have you fixing all this technical debt.

**Comment by matthewmturner** [comment]
Happy to help!

> 
> In any case, we should be calling pytest with `not hdfs`, and that should prevent `test_filesystems` from running.

Should I be doing something for the above?

I see the below in `.github/main.yaml` already (already calling `not hdfs`) but are you referring to something else?


```
name: run tests
      run: PYTEST_BACKENDS=$BACKENDS PYTEST_EXPRESSION="not hdfs" ./ci/run_tests.sh
      shell: bash
```

Okay I'll create an issue and new PR for moving the imports out of global namespace.

**Comment by datapythonista** [comment]
> I see the below in `.github/main.yaml` already (already calling `not hdfs`) but are you referring to something else?

That's it. The problem is that even if the tests marked with `hdfs` are skipped, the test files are loaded (e.g. `tests/test_filesystem.py`) and anything in the global namespace would still be executed. So, any import may end up triggering other modules, and the impyla module may be still loading and failing.

**Comment by matthewmturner** [comment]
@datapythonista Pandas / files CI is failing on `test_sqlalchemy.py` with `from ibis.tests.sql.test_compiler import ExprTestCases` which is used here:

```
class TestSQLAlchemySelect(unittest.TestCase, ExprTestCases):
    def setUp(self):
        self.con = MockAlchemyConnection()
        self.alltypes = self.con.table('functional_alltypes')
        self.sa_alltypes = self.con.meta.tables['functional_alltypes']
        self.meta = sa.MetaData()

        self.sa_star1 = self._get_sqla('star1')
```

its not clear to me how to move the import into the class in this case like we did on the last because the class is inheriting from it.  im also confused because this fails with same error on the SQL tests - which i wouldnt expect. I had thought the IbisTestEnv update we did (which was merged into master) would fix this - was i misunderstanding / what are your thoughts?

**Comment by datapythonista** [comment]
I don't understand that `test_compiler` file, seems to be mixing things. There are few things going on. That file seems to be skipped for no reason, see #2477. Then there is some weird stuff with tests, see #2488. And finally, that file is in `sql` (which soon will become `backends/base_sqlalchemy`), and my understanding is that SQLAlchemy implements its own SQL compiler, and the backends shouldn't be depending on Impala (for sure) or depend on the newly `base_sql` backend either.

I think we'll have to have a look at those three issues, and do some more refactoring, if those problems are affecting this PR (it shouldn't, be looks like they are).

**Comment by matthewmturner** [comment]
Looks like were back to the error in `test_compiler.py`. going to try picking that one back up.  let me know if you think there is something else going on.

**Comment by datapythonista** [comment]
Yes, looks like the error will be fixed with #2485. We're almost there, really great job decoupling things.

**Comment by datapythonista** [comment]
@matthewmturner when you have time can you merge master and see if #2501 has fixed the tests failures please?

**Comment by matthewmturner** [comment]
i think we need to refactor `ibis/__init__.py` since previously packages required impala and there was a specific check for that.

```
def __getattr__(name):
        if name in ('HDFS', 'WebHDFS', 'hdfs_connect'):
            warnings.warn(
                f'`ibis.{name}` has been deprecated and will be removed in a '
                f'future version, use `ibis.impala.{name}` instead',
                FutureWarning,
                stacklevel=2,
            )
            if 'impala' in globals():
                return getattr(impala, name)
            else:
                raise AttributeError(
                    f'`ibis.{name}` requires impala backend to be installed'
                )
        raise AttributeError
```

besides that looking pretty good. had to make a few subsequent tweaks but looking much better in the aggregate.

@datapythonista  do you have thoughts on how to handle the above?

**Comment by datapythonista** [comment]
> @datapythonista do you have thoughts on how to handle the above?

This is weird. the first `if` should only be executed if any of the `hdfs` functions is called (e.g. `impala.HDFS`). And that should only happen in the Impala backend.

Not sure what is wrong, but from the traceback seems like `name` is `pyspark`, meaning that `ibis.pyspark` is the call that triggered this call to `__getattr__`. But `if name in ('HDFS', 'WebHDFS', 'hdfs_connect'):` should be false, and never enter the `if`. I'm missing something.

**Comment by datapythonista** [comment]
I see the error in the docs is saying that the `hdfs_connect` function does not exist.
```
>>>-------------------------------------------------------------------------
Exception in /ibis/docs/source/backends/impala.rst at block ending on line 218
Specify :okexcept: as an option in the ipython:: block to suppress this message
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-3-e949c05499fe> in <module>
----> 1 hdfs = ibis.impala.hdfs_connect(host=host)

AttributeError: module 'ibis.impala' has no attribute 'hdfs_connect'
<<<-------------------------------------------------------------------------
```
Seems related to the error in the Spark build. Let me know if you can't find the issue, I'll have a look.

File: ibis/backends/spark/ddl.py
Comment (datapythonista) [inline ibis/backends/spark/ddl.py:3]: This seems to be the problem. We are still importing impala from spark. Whatever it's used in Spark should be moved to `base_sql`, and not import impala at all. Otherwise spark will load the impala backend, and will fail because it's dependencies won't be installed.

This should be fixed in a separate PR. Do you want to work on it? I can work on it if you prefer.

Comment (matthewmturner) [reply ibis/backends/spark/ddl.py:3]: submitted a new PR #2513

**Comment by jreback** [comment]
can u merge master

**Comment by matthewmturner** [comment]
@datapythonista any thoughts on the test error for import time?

**Comment by datapythonista** [comment]
That happens randomly, will have a look later. The one we need to fix is the documentation. Seems like something has been changed, I guess we are exposing internal objects of the backend in the docs, and their module will have to be updated: https://dev.azure.com/ibis-project/ibis/_build/results?buildId=3797&view=logs&j=8f09edc2-e3b7-52de-126a-0225c4f3efa1&t=e3ba5a33-21a6-5484-d0be-59baf122f79c&l=510

File: ibis/backends/base_sql/__init__.py
Comment (datapythonista) [inline ibis/backends/base_sql/__init__.py:21]: We shouldn't have imports from the base backends to the specific ones, only the other way round. Can you move this `identifiers` to the base backend please?

File: ibis/backends/impala/compiler.py
Comment (datapythonista) [inline ibis/backends/impala/compiler.py:18]: Can you remove this? I guess it was left from tests.

**Review by datapythonista — COMMENTED** [review]
Thanks @matthewmturner looks great. Just couple of comments. And also, looks like you committed some binary files from the doc build, if you can please remove them.

But almost ready, really happy to see this finally merged, thanks for the huge effort.

**Review by jreback — CHANGES_REQUESTED** [review]
just @datapythonista comments

**Comment by datapythonista** [comment]
Looks like there is a failing test.

**Comment by matthewmturner** [comment]
> 
> 
> Looks like there is a failing test.

I made some more updates.  But i think there will be an issue with an `ImpalaE2E` being imported into kudu tests.  im not really sure what that is and it doesnt appear to exist nor does the module that it comes from (`impala.common`).

**Comment by matthewmturner** [comment]
@datapythonista azp timed out - can you rerun?

**Comment by matthewmturner** [comment]
Does consecutive azp timeouts indicate there could be something wrong in the code?

**Comment by matthewmturner** [comment]
@datapythonista @jreback anything we can do here to fix the azp timeout besides rerunning?

**Comment by datapythonista** [comment]
Not sure what's going on. The azure log wasn't working for me, but I guess it's conda. We shouldn't be so far from moving those builds to actions, so I don't want to spend time on them. Maybe you can push an empty commit, to see if there is any caxlche or something making things when rerunning?

**Comment by matthewmturner** [comment]
@datapythonista @jreback all green!

**Comment by jreback** [comment]
the problem here is conda is taking 50m which is completely crazy

the recipie needs to change

ok for this PR but we need to trim things down before the next release as this makes the CI extremely unstable

**Comment by jreback** [comment]
i don't think moving the builds to actions makes any difference 

the problem is the environment file has everything in it
it needs to just be ibis with no deps

yea the docs will have to be dealt with but likely you only need minimal backends for that

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @matthewmturner

**Pull request merged**
